### PR TITLE
perf: Add init and cleanup of long lived resources

### PR DESCRIPTION
**Pull request by robhowley**
# What this PR does / why we need it:

* add `initialize` and `close` to `FeatureStore`, `Provider`, `OnlineStore` ~~and `OfflineStore`~~
* they are async and default impl does nothing
* provides a hook for datastores like dynamodb which needs to do some setup in order to create reusable clients and sessions
* added the dynamo setup

_update_: added `pytest-asyncio` to ci which is +555 of the diff

# Which issue(s) this PR fixes:

Addresses the performance penalty mentioned [here](https://github.com/feast-dev/feast/pull/4244#discussion_r1623274405) by @tokoko 

# Misc

### CODE REVIEW

File: sdk/python/feast/feature_server.py [hunk 1]
```diff
@@ -100,8 +100,10 @@ def async_refresh():
     @asynccontextmanager
     async def lifespan(app: FastAPI):
         async_refresh()
+        await store.initialize()
         yield
         stop_refresh()
+        await store.close()
```
Comment (robhowley) [inline sdk/python/feast/feature_server.py:106]: can create and cleanup resources as part of the feature server lifespan

**Comment by robhowley** [comment]
not 100% what to make this error in the py39 unit tests

```
ERROR sdk/python/tests/unit/permissions/auth/server/test_auth_registry_server.py::test_registry_apis[\nauth:\n  type: no_auth\n-applied_permissions0] - RuntimeError: Failed to bind to address [::]:34361; set GRPC_VERBOSITY=debug environment variable to see detailed error message.
```

**Comment by franciscojavierarceo** [comment]
@dmartinol @redhatHameed  could you take a look?

**Comment by dmartinol** [comment]
> not 100% what to make this error in the py39 unit tests
> 
> ```
> ERROR sdk/python/tests/unit/permissions/auth/server/test_auth_registry_server.py::test_registry_apis[\nauth:\n  type: no_auth\n-applied_permissions0] - RuntimeError: Failed to bind to address [::]:34361; set GRPC_VERBOSITY=debug environment variable to see detailed error message.
> ```

I cloned your repo and it succeeded with no issues, maybe a temporary env issue?

**Comment by dmartinol** [comment]
Some questions about this change:
- why adding the init/close methods to `offline_store` if there is no implementation yet? 
- if you really think that `offline_store` deserves them, you probably need to invoke them on the `store` instance in `offline_store` module as well. Probably the `feature_server`, which actually is more an `online_server`, will not have an `offline_store` at all, or may use a remote one.

@franciscojavierarceo @tokoko

**Comment by robhowley** [comment]
> Some questions about this change:
> 
> * why adding the init/close methods to `offline_store` if there is no implementation yet?
> * if you really think that `offline_store` deserves them, you probably need to invoke them on the `store` instance in `offline_store` module as well. Probably the `feature_server`, which actually is more an `online_server`, will not have an `offline_store` at all, or may use a remote one.
> 
> @franciscojavierarceo @tokoko

yea that's a good point. offline store doesn't make a ton of sense in the server case. was just trying to maintain consistency across the two, but that need not be the case. i'll remove.

**Comment by tokoko** [comment]
@robhowley thanks, I'm not too sure about the changes though. can you elaborate on why we need this?

Is the client creation actually that expensive that it needs a separate init method? I think what I was pointing out in the comment you linked above was that `create_client` was being called on every invocation rather than only during the first one and cached afterwards. The first one being slightly slower than others doesn't seem that bad to me.

I get why a separate initializer can be beneficial, but I'm a little uneasy that it pushes us into the mixture of sync/async methods. What if I'm invoking the online read the old-fashioned way w/o asyncio, do I still need to call async initializer first or is that only relevant for async calls?

The same goes for the close method. This has actually come up before in #4401. As I pointed out there, I'm not sure what closing an online store should actually mean. Do calls on a closed online store error out or do they still succeed, but need to recreate the resources? I'd rather not have to answer those questions at all unless you think we absolutely need to.

**Comment by robhowley** [comment]
With a little time to mull this over, I'm curious what everyone's thoughts are on this. I'm not attached to the open/close idea, but it does seem like there should be a way to clean up resources. Most of async support will come from libraries expecting to be used in an aync context manager, so some kind of equivalent mechanism should be available.

**Comment by tokoko** [comment]
@robhowley how would the normal non-async workflow be affected with this? If we put connection creation in an async initializer that won't be called by a non-async user... would we implicitly run an initializer anyway during the first call?

**Comment by robhowley** [comment]
The sync usage shouldn't be impacted. The async code paths need not be touched by the sync code, so I don't think any async initialization would be needed/called.

```
# in sdk mode, this shouldn't touch async code
# and as a result shouldn't require any async connections to be created
feature_df = FeatureStore().get_online_features(entities).to_df()
```

**Comment by robhowley** [comment]
@franciscojavierarceo @tokoko im not exactly sure what to make of the failing integration tests. created an issues #4658 bc they don't seem to have passed in better part of two months. im assuming you're aware, just (1) thought it worth flagging and (2) curious if you had any opinions on how concerned i should be about it

**Comment by robhowley** [comment]
@tokoko @franciscojavierarceo btw, some added context on this mr ... i've been running this in prod for a few days and reusing the client reduces event loop lag in the server since creation of the client is a blocking operation. the scalability and speed improvements are obviously quite good as a result. reusing the client, however, should be paired w the ability to clean up the resources.

**Comment by robhowley** [comment]
would it perhaps be clearer to just go full async context manager? somewhat more explicit that this isnt going to impact the sync code paths?

File: sdk/python/feast/feature_server.py [hunk 2]
```diff
@@ -101,8 +101,10 @@ def async_refresh():
     @asynccontextmanager
     async def lifespan(app: FastAPI):
         async_refresh()
+        await store.initialize()
```
Comment (franciscojavierarceo) [inline sdk/python/feast/feature_server.py:104]: I wonder if this is the root cause

Comment (robhowley) [reply sdk/python/feast/feature_server.py:104]: root cause of what?

File: sdk/python/feast/feature_server.py (hunk 1 above)
Comment (franciscojavierarceo) [reply sdk/python/feast/feature_server.py:107]: Can you remove this and let's rerun the integration test? Or maybe create a separate branch without it?

File: sdk/python/feast/feature_server.py (hunk 2 above)
Comment (robhowley) [reply sdk/python/feast/feature_server.py]: for all but dyanmo this method call is just `pass`

Comment (robhowley) [reply sdk/python/feast/feature_server.py]: it failed integration tests [bc the dynamo client is null](https://github.com/feast-dev/feast/actions/runs/11486758859/job/31969747403?pr=4642#step:15:2352) during the test using the server

Comment (franciscojavierarceo) [reply sdk/python/feast/feature_server.py:104]: i should have been more explicit, shouldn't initialization happen before the refresh? I took a quick look at the code but wasn't sure.

Comment (robhowley) [reply sdk/python/feast/feature_server.py:104]: they don't really interact w each other. does read like initialize should be first, but it only creates async clients, doesn't deal w the registry

**Comment by robhowley** [comment]
@franciscojavierarceo just a heads up, [the remaining integration test failure](https://github.com/feast-dev/feast/actions/runs/11488391804/job/31975025552?pr=4642#step:15:907) comes from [here](https://github.com/feast-dev/feast/blob/master/sdk/python/tests/integration/online_store/test_python_feature_server.py#L61). the push works, but the subsequent retrieval to confirm it's there, in the case of dynamo, is async. this is a sync function so there's no running event loop. working on a fix to get this green

File: sdk/python/feast/feature_server.py (hunk 2 above)
Comment (robhowley) [reply sdk/python/feast/feature_server.py:103]: swapped the order, does read better that way

File: sdk/python/tests/integration/online_store/test_python_feature_server.py [hunk 3]
```diff
@@ -137,5 +135,5 @@ def python_fs_client(environment, universal_data_sources, request):
     feast_objects.extend([driver(), customer(), location()])
     fs.apply(feast_objects)
     fs.materialize(environment.start_date, environment.end_date)
-    client = TestClient(get_app(fs))
-    yield client
+    with TestClient(get_app(fs)) as client:
+        yield client
```
Comment (robhowley) [inline sdk/python/tests/integration/online_store/test_python_feature_server.py:139]: [using test client as a context manager allows fastapi to incorporate lifespan](https://fastapi.tiangolo.com/advanced/testing-events/). dynamo needs this bc the connection and by extention event loop management happens at startup/shutdown. this fixed the remaining integration test issues.

File: sdk/python/pytest.ini [hunk 4]
```diff
@@ -7,6 +9,8 @@ env =
     IS_TEST=True
 
 filterwarnings =
+    error::_pytest.warning_types.PytestConfigWarning
+    error::_pytest.warning_types.PytestUnhandledCoroutineWarning
```
Comment (robhowley) [inline sdk/python/pytest.ini:13]: fail if any async test functions are skipped bc of missing plugins

**Review by franciscojavierarceo — APPROVED** [review]
Thank you for working through this and for shipping it!
image

**Pull request merged**